src/compiler/checker.ts

```diff
@@ -29246,6 +29247,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     function isPossiblyDiscriminantValue(node: Expression): boolean {
         switch (node.kind) {
             case SyntaxKind.StringLiteral:
+                return !isEmptyStringLiteral(node as StringLiteral);

```

Comment:
I think that it's likely that we could fix this in a more general way if we'd do **something** like this here:
```suggestion
                return !hasSkipDirectInferenceFlag(node);
```

The problem is that only nodes within a `containingCall` are "blocked"/skipped today:
https://github.dev/microsoft/TypeScript/blob/a280cafbf85df16bcd2f7258f26fdb85615b3c41/src/compiler/checker.ts#L1851-L1858

Maybe it's just a matter of adding a new stop condition to this ancestry traversal?

Reply:
Yeah, I couldn't resist tinkering with this a little bit more and I think that this kind of solution is on the right track.

With this patch the expanded test case works:
```diff
-        const containingCall = findAncestor(node, isCallLikeExpression);
+        const containingCall = findAncestor(node, n => isCallLikeExpression(n) || !!n.parent && isVariableDeclaration(n.parent));
```

I used this condition only as a quick test to see if it could work. By no means it's correct/complete. I think that we are looking for a parent that "stops" providing contextual types as we should mark everything up until that parent. Since contextual types come from lots of places... I'm not sure how to write such a condition on the spot :p

Reply:
> @Andarist the idea I have in mind is doing a smarter filtering [here](https://insiders.vscode.dev/github/MariaSolOs/TypeScript/blob/2f75e38bc40d0c58ab74ba8cd4ba0dcad33ce4f3/src/compiler/checker.ts#L29271-L29275) and filter out properties where the initializer doesn't match the property's type.

That makes a lot of sense 👍